cluster-autoscaler/cloudprovider/mcm/mcm_manager.go

```diff
@@ -798,21 +859,87 @@ func isRollingUpdateFinished(md *v1alpha1.MachineDeployment) bool {
 
 // getMachineDeploymentUntilDeadline returns error only when fetching the machineDeployment has been failing consequently and deadline is crossed
 func (m *McmManager) getMachineDeploymentUntilDeadline(mdName string, retryInterval time.Duration, deadline time.Time) (*v1alpha1.MachineDeployment, error) {
+	ctx, cancelFn := context.WithDeadline(context.Background(), deadline)
+	defer cancelFn()
 	for {
 		md, err := m.machineDeploymentLister.MachineDeployments(m.namespace).Get(mdName)
-		if err != nil && time.Now().Before(deadline) {
-			klog.Warningf("Unable to fetch MachineDeployment object %s, Error: %s, will retry in %s", mdName, err, retryInterval)
-			time.Sleep(conflictRetryInterval)
-			continue
-		} else if err != nil {
-			// Timeout occurred
-			klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %s, timeout occurred", mdName, err)
-			return nil, err
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %+v, timeout occurred", mdName, err)
+				return nil, err
+			case <-time.After(retryInterval):
+				klog.Warningf("Unable to fetch MachineDeployment object %s, Error: %+v, will retry the operation", mdName, err)
+				continue
+			}
 		}
 		return md, nil
 	}
 }
 
+// getMachineUntilDeadline returns error only when fetching the machine has been failing consequently and deadline is crossed
+func (m *McmManager) getMachineUntilDeadline(mcName string, retryInterval time.Duration, deadline time.Time) (*v1alpha1.Machine, error) {
+	ctx, cancelFn := context.WithDeadline(context.Background(), deadline)
+	defer cancelFn()
+	for {
+		mc, err := m.machineLister.Machines(m.namespace).Get(mcName)
+		if err != nil {
+			select {
+			case <-ctx.Done():
+				klog.Errorf("Unable to fetch MachineDeployment object %s, Error: %+v, timeout occurred", mcName, err)
+				return nil, err
+			case <-time.After(retryInterval):
+				klog.Warningf("Unable to fetch MachineDeployment object %s, Error: %+v, will retry the operation", mcName, err)
+				continue
+			}
+		}
+		return mc, nil
+	}
+}
+
+// getMachinesForMachineDeploymentUntilDeadline returns error only when fetching the machine list has been failing consequently and deadline is crossed
+func (m *McmManager) getMachinesForMachineDeploymentUntilDeadline(mdName string, retryInterval time.Duration, deadline time.Time) (*v1alpha1.MachineList, error) {
+	ctx, cancelFn := context.WithDeadline(context.Background(), deadline)
+	defer cancelFn()
+	for {
+		ml, err := m.machineClient.Machines(m.namespace).List(ctx, metav1.ListOptions{LabelSelector: labels.FormatLabels(map[string]string{machineDeploymentNameLabel: mdName})})
```

Comment:
use a lister call instead of api server call , if possible here. We should use api server calls only for get of machinedeployment and `update` of anything.

Reply:
addressed in [7824ca9](https://github.com/gardener/autoscaler/pull/215/commits/7824ca90fa42fba9b53f09ec7d7d90c923b3b7b3)